### PR TITLE
Port SAC to ray 0.6.6

### PR DESCRIPTION
**Pull request by AdityaGudimella**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            if get_policy_class is None:
+                policy = default_policy
+            else:
+                policy = get_policy_class(config)
+            if before_init:
+                before_init(self)
+            if make_workers:
+                self.workers = make_workers(self, env_creator, policy, config)
+            else:
+                self.workers = self._make_workers(env_creator, policy, config,
```
[the remark this answers is not in the document]
Comment (RuofanKong) [reply python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py]: BTW, looks _make_worker() is missing if I'm correct.

File: python/ray/rllib/agents/sac/sac/rllib_proxy/_tf_policy_template.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            def before_loss_init_wrapper(policy, obs_space, action_space,
+                                         config):
+                if before_loss_init:
+                    before_loss_init(policy, obs_space, action_space, config)
+                if extra_action_fetches_fn is None:
+                    self._extra_action_fetches = {}
+                else:
+                    self._extra_action_fetches = extra_action_fetches_fn(self)
+
+            DynamicTFPolicy.__init__(
```
Comment (RuofanKong) [inline python/ray/rllib/agents/sac/sac/rllib_proxy/_tf_policy_template.py]: Ray 0.6.6 doesn't have the class `DynamicTFPolicy`.

File: python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        def _init(self, config, env_creator):
+            if validate_config:
+                validate_config(config)
+            if get_initial_state:
+                self.state = get_initial_state(self)
+            else:
+                self.state = {}
+            if get_policy_class is None:
+                policy = default_policy
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py]: `BonsaiSACTrainer = SACTrainer.with_updates(default_policy=BonsaiSACPolicy)`

File: python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            else:
+                self.workers = self._make_workers(env_creator, policy, config,
+                                                  self.config["num_workers"])
+            if make_policy_optimizer:
+                self.optimizer = make_policy_optimizer(self.workers, config)
+            else:
+                optimizer_config = dict(
+                    config["optimizer"],
+                    **{"train_batch_size": config["train_batch_size"]})
+                self.optimizer = SyncSamplesOptimizer(self.workers,
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/sac/rllib_proxy/_trainer_template.py]: Same as above

File: python/ray/rllib/agents/sac/sac/custom_model.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+import tensorflow as tf
+
+from sac import config
+from sac.rllib_proxy import TFModelV2, normc_initializer, get_activation_fn
+
+ArrayShape = Tuple[int, ...]
+KernelInitializer = Callable[[], tf.keras.initializers.Initializer]
+
+
+def build_fcn(
```
Comment (RuofanKong) [inline python/ray/rllib/agents/sac/sac/custom_model.py]: Question - So later on when we support Luminance, we need to have a similar one `def build_convn()` right?

File: python/ray/rllib/agents/sac/tests/test_sac.py [hunk 6]
```diff
@@ -49,12 +54,172 @@ def trainer(config):
 @pytest.mark.usefixtures("ray_env")
 def test_convergence(trainer, env_id, threshold):
     mean_episode_reward = -float("inf")
-    for i in range(300):
+    for i in range(350):
```
Comment (RuofanKong) [inline python/ray/rllib/agents/sac/tests/test_sac.py]: How about not changing to 350?

**Comment by AdityaGudimella** [comment]
> Is there any reason to do not have the vf_share_layers option in dcm?
> Why do you not include model_config?

The `build_fcn` function is supposed to be much more simpler than what RLLib is trying to do with their `FCNet` class. The FCNet class is aware of details like actor/critic architectures and so on. `build_fcn` on the other hand is a much more simpler class that just builds an FCN according to a given (opinionated) specification. This means that `build_fcn` can be used internally within the FCNet class if needed. It wasn't meant to be a replacement to the FCNet class.

As far as the model config goes, as I said above, the `build_fcn` function is meant to be as simple and opinionated as possible. It is merely a utility function. Whoever is using the function (which, again, can be the FCNet class), will parse any config and pass the required arguments to the `build_fcn` function. In the case of RLLib, it is the SACModel class that does this.

**Comment by AdityaGudimella** [comment]
> `CustomModel` is not used any more, Can we Remove it?

Done

File: python/ray/rllib/agents/sac/tests/test_sac.py (hunk 6 above)
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/tests/test_sac.py]: Done

File: python/ray/rllib/agents/sac/sac/custom_model.py (hunk 5 above)
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/sac/custom_model.py]: For SAC, yes.

File: python/ray/rllib/agents/sac/sac/custom_model.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        activation=None,
+        kernel_initializer=normc_initializer(0.01),
+    )(last_layer)
+
+    reshaped_value_out = tf.keras.layers.Reshape([-1])(value_out)
+
+    return tf.keras.Model(inputs, [layer_out, reshaped_value_out], name=name)
+
+
+class CustomModel(TFModelV2):
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/sac/custom_model.py]: Removed

File: python/ray/rllib/agents/sac/sac/custom_model.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    output = tf.keras.layers.Dense(
+        num_outputs,
+        name="fc_out",
+        activation=output_activation,
+        kernel_initializer=kernel_initializers[-1]()
+    )(last_layer)
+    return tf.keras.Model(inputs, [output], name=name)
+
+
+def build_model(
```
[the remark this answers is not in the document]
Comment (AdityaGudimella) [reply python/ray/rllib/agents/sac/sac/custom_model.py]: Removed

**Pull request closed without merging**